core/src/main/java/com/google/cloud/sql/core/CoreSocketFactory.java

```diff
@@ -220,6 +224,40 @@ public static Socket connect(Properties props, String unixPathSuffix) throws IOE
     return getInstance().createSslSocket(csqlInstanceName, ipTypes);
   }
 
+  /**
+   * Returns data that can be used to establish Cloud SQL SSL connection.
+   */
+  public static SslData getSslData(Properties props) {
+    final String csqlInstanceName = props.getProperty(CLOUD_SQL_INSTANCE_PROPERTY);
+
+    return getInstance().computeSslData(csqlInstanceName);
+  }
+
+  private SslData computeSslData(String instanceName) {
+    CloudSqlInstance instance =
+        instances.computeIfAbsent(
+            instanceName, k -> new CloudSqlInstance(k, adminApi, executor, localKeyPair));
+    return instance.getSslData();
+  }
+
+  /**
+   * Returns preferred ip address that can be used to establish Cloud SQL connection.
+   */
+  public static String getHostIp(Properties props) {
+    final String csqlInstanceName = props.getProperty(CLOUD_SQL_INSTANCE_PROPERTY);
+    final List<String> ipTypes = listIpTypes(props.getProperty("ipTypes", DEFAULT_IP_TYPES));
+
+    return getInstance().getHostIp(csqlInstanceName, ipTypes);
+  }
+
+  private String getHostIp(String instanceName, List<String> ipTypes) {
+    CloudSqlInstance instance =
```

Comment:
nit: This `instances.computeIfAbsent(instanceName, k -> new CloudSqlInstance(k, adminApi, executor, localKeyPair));` is being used in a couple of different places now - would it make sense to move it into it's own function to avoid any instances accidentally getting created differently?